server/lib/compareUserPasswordHistory.ts

```diff
@@ -0,0 +1,46 @@
+import { Accounts } from 'meteor/accounts-base';
+
+import { IUser } from '../../definition/IUser';
+import { IPassword } from '../../definition/IPassword';
+import { settings } from '../../app/settings/server';
+
+/**
+ * Check if a given password is the one user by given user or if the user doesn't have a password
+ * @param {object} user User object
+ * @param {object} pass Object with { plain: 'plain-test-password' } or { sha256: 'sha256password' }
+ */
+export function compareUserPasswordHistory(user: IUser, pass: IPassword): boolean {
+	if (!user?.services?.passwordHistory || !settings.get('Accounts_Password_History_Enabled')) {
+		return true;
+	}
+
+	if (!pass || (!pass.plain && !pass.sha256) || !user?.services?.password?.bcrypt) {
+		return false;
+	}
+
+	const currentPassword = user.services.password.bcrypt;
+	const passwordHistory = user.services.passwordHistory.slice(-Number(settings.get('Accounts_Password_History_Amount')));
+
+	for (const password of passwordHistory) {
+		if (!password.trim()) {
+			user.services.password.bcrypt = currentPassword;
+			return false;
+		}
+		user.services.password.bcrypt = password;
+
+		const historyPassword = pass.plain || {
+			digest: pass.sha256 ? pass.sha256.toLowerCase() : '',
+			algorithm: 'sha-256' as const,
+		};
+
+		const passCheck = Accounts._checkPassword(user, historyPassword);
+
+		if (!passCheck.error) {
+			user.services.password.bcrypt = currentPassword;
+			return false;
+		}
+	}
+
+	user.services.password.bcrypt = currentPassword;
```

Comment:
What is this line for? I noticed you used it 3 times in here 👀

Reply:
In this function, I use the `_checkPassword` method (which requires a `Meteor.User` object) for every entry of the password history -- in order to compare the passwords in the history with the new password. Instead of creating an empty `Meteor.User` object for each password in the history, I decided to use the `user` object that is received by this function, and temporarily assign each password in the history to it.

So, before returning the value, I must assign the original password back to the `user` (so that this object is not changed in the caller function too).

I first intended to create a local `IUser` object, but it hasn't worked with the `_checkPassword` method. That's why I used the `user` param. Do you think that's ok or do you have any suggestions?

Reply:
If that's the case, I'd have preffered separating the code that performed the actual password check on `checkPassword` to its own function, so you can call it with 2 passwords instead of modifying the object each time you wanted to do a comparison 👀 what do you think?

Reply:
Ah, realizing it may not be possible right? `_checkPassword` is hidden in meteor magic so :rip:. Ignore my recommendation haha